add methods to `PudlTabl` so it can be serialized and de-serialized

arengel opened the pull request:
## Motivation

The purpose of this PR is to enable `PudlTabl` to be serialized using `pickle`. In RMI's work with PUDL, we've often found it useful to save a copy of a `PudlTabl` before testing or running our data processing. This can save substantial amounts of time without jeopardizing validity because essentially all of our code is downstream of PUDL. The proposed changes here are a refined and generalized version of the solution we have used for a while now. In addition to being a more elegant solution than us subclassing or wrapping `PudlTabl`, it bring this functionality to all users, and the proposed tests should ensure that it continues to work properly.

## The changes

1. Two new methods for `PudlTabl`. A description of the default versions of these methods can be found [here](https://docs.python.org/3/library/pickle.html#pickle-inst), an example of how they might be modified in special cases is discussed [here](https://docs.python.org/3/library/pickle.html#pickle-state).
   1. `__getstate__` this converts the `PudlTabl` state into one that can be serialized with `pickle`. This is required because `pickle` doesn't like the `lambda` in the `self._dfs` `defaultdict` or `sa.Engine` so they need to be converted into form that `pickle` can deal with.
   2. `__setstate__` this takes the state as it was saved above and puts the pieces back into `PudlTabl` so a restored `PudlTabl` behaves the same as it would have before it was serialized. One important detail is that ``__init__`` is not run when an  object is de-serialized, so under some circumstances, setup that happened there may also need to occur here.

 2. Two new tests 
    1. `test_pudltabl_pickle` - this tests that we can take a `PudlTabl` with at least one df and that it can be pickled and restored from a pickle.
    2. `test_pudltabl_datazip` - this is similar to the previous but instead of using pickle, uses `DataZip` from our [`etoolbox`](https://github.com/RMI/etoolbox) which uses the same protocol as `pickle` (as far as I understand it). We generally use DataZip rather that pickle because it creates much more compact files because it uses `parquet` for dataframes and so is also more portable than pickle. It is here more as an example since I am not proposing to add a dependency to PUDL. Currently the test is skipped if `etoolbox` is not installed. Happy to remove it, chat about DataZip, or both.

## A few additional points

1. The current approach works and can accommodate the `Datastore` object because there is no issue with pickling the `__dict__` attribute of one of its instances. Not fully understanding how `Datastore` works, I think that's alright but am not confident of that. (`Datastore` does break `DataZip` so `DataZip` knows to exclude it, this has been fine for our purposes but would probably need to be addressed if you wanted to officially support `DataZip`. And to be clear, I am not here proposing that.)
2. My approach to saving the state of the `pudl_engine` is to save the URL, and on restoring, using `sa.create_engine` to create a new `sa.Engine` from the saved URL. This works fine. But then I thought, that will produce a `PudlTabl` that won't work properly if it is moved from the user/machine on which it was created. To fix this I added a check that tries connecting to the engine created from the saved URL, if that fails, we create an engine from the URL in `pudl.workspace.setup.get_defaults`. While that is probably convenient, it might cause weird problems I haven't thought of.
3. Both tests use the fixture `fast_out_annual`. This works totally fine when running the whole test suite but creating that fixture just to test serializing `PudlTabl` is overkill. For now this seemed like the better approach than creating another `PudlTabl`, especially since I wasn't sure what DB to connect it to so that it would work in GHA/CI land. If more work is required, especially as relates to the point below, it might be worth creating a minimal `PudlTabl` fixture for testing this functionality. Here I would request guidance on how to set it up so it plays nicely.
4. Again on the test, not fully understanding `Datastore` I'm not sure how to test that the restored one is functioning properly. For our current use case this isn't really a problem but could be for other users or in the future. I'd be happy to work on addressing this, though may request some guidance on how to think about `Datastore`.


## PR Checklist

- [x] Merge the most recent version of the branch you are merging into (probably `dev`).
- [x] All CI checks are passing. [Run tests locally to debug failures](https://catalystcoop-pudl.readthedocs.io/en/latest/dev/testing.html#running-tests-with-tox)
- [x] Make sure you've included good docstrings.
- [ ] For major data coverage & analysis changes, [run data validation tests](https://catalystcoop-pudl.readthedocs.io/en/latest/dev/testing.html#data-validation)
- [x] Include unit tests for new functions and classes.
- [ ] Defensive data quality/sanity checks in analyses & data processing functions.
- [x] Update the [release notes](https://catalystcoop-pudl.readthedocs.io/en/latest/release_notes.html) and reference reference the PR and related issues.
- [x] Do your own explanatory review of the PR to help the reviewer understand what's going on and identify issues preemptively.

jdangerx commented on src/pudl/output/pudltabl.py:
```diff
@@ -1207,6 +1207,62 @@ def epacamd_eia(
+        """
+        try:
+            pudl_engine = sa.create_engine(state["pudl_engine"])
+            # make sure that the URL for the engine from ``state`` is usable now, if
```
I think serializing sqlalchemy to a DB connection URL is pretty clever!

Regarding the fallback behavior: typically, I think it's important to have your serialization/deserialization be inverse operations of each other, and this breaks that guarantee. However, I can see how this would be very ergonomic in some cases, like passing pickled objects back and forth across different users. I think we can keep the fallback behavior, but we should alert the user with some sort of log, so that the user can see "Oh, I'm not getting *exactly* the same object back."

arengel replied:
That's a great point, I'll add a logger message to that effect.

arengel replied:
Another thought, should something like this happen with `Datastore.local_cache_path`?

jdangerx commented on src/pudl/output/pudltabl.py:
```diff
@@ -1207,6 +1207,62 @@ def epacamd_eia(
+        This method is run when the object is restored from a pickle. Anything
+        that was changed in :meth:`pudl.output.pudltabl.PudlTabl.__getstate__` must be
+        undone here. Another important detail is that ``__init__`` is not run when an
+        object is de-serialized, so under some circumstances, setup that happened there
```
What circumstances are these?

arengel replied:
I think usually circumstances where `__init__` does things outside the class like changing globals, opening a file, etc. I altered the text to be less mysterious.

 I put this comment for two reasons (1) for anyone looking at this method that wasn't familiar with how pickle restores objects because it's not immediately obvious how this fits into the rest that process, and (2) in case something special needs to be done for `Datastore`. Granted (2) is sort of redundant given that is better and more clearly addressed by the PR narrative. I think I was just a little paranoid about breaking `Datastore` resulting in odd behavior.

jdangerx commented on test/integration/output_test.py:
```diff
@@ -252,3 +252,51 @@ def test_mcoe_filled(fast_out_filled, df_name, expected_nuke_fraction, tolerance
+
+    If etoolbox is installed, test that PudlTabl can be serialized with DataZip.
+    """
+    dz = pytest.importorskip("etoolbox.datazip")
```
`etoolbox` isn't in PUDL requirements, so this will always be a skip in our tests. I'm kind of inclined to say we should keep this downstream, so we don't have a confusing "always skip because this doesn't actually apply to our code as is" test in our codebase.

arengel replied:
Totally fair, I'll remove the test.

jdangerx reviewed (changes requested):
Hello! Thanks for the contribution and sorry for being a bit slow on the review - I've been doing some holiday/family stuff this weekend.

Overall this looks good, and the tests make sense too :) I think there are a few places where we would want to make things a little clearer for other users.

Finally, there's some github actions authentication stuff that I need to work out, which isn't your fault. I'll pull that to the top of the queue for the next sprint, but until then you won't be able to merge this PR - sorry about that!

arengel commented:
Thanks for the thoughtful review @jdangerx ! I believe I've addressed the requested changes by (1) adding a logger message if the default DB URL must be used and (2) removing the test that would always skip. I also changed the somewhat cryptic `__setstate__` doctstring message about when it's important to realize that `__init__` does not get run when an object gets de-serialized. I could also see removing that comment entirely since that docstring is maybe not the place to discuss how pickle works.

I realize merging must wait on the GitHub Action issue but want to make sure this PR is otherwise ready to go.

jdangerx reviewed (approved):
Thanks for the quick updates :) I'll look into the actions stuff in the next sprint. @cmgosnell -  @bendnorman mentioned you might have run into some "CI breaks for fork-based PRs" stuff in the past, so flagging this in case you have any insight.

bendnorman commented:
The GitHub actions are failing because PRs from forks branches don't have access to PUDL's GitHub secrets. @jdangerx is working on updating our actions so they don't need access to GitHub secrets. If you need this work in the `dev` branch ASAP @arengel, you are listed as a collaborator on the PUDL repo so you can create a new PR from a branch inside this repo so the actions will work.

arengel commented:
> The GitHub actions are failing because PRs from forks branches don't have access to PUDL's GitHub secrets. @jdangerx is working on updating our actions so they don't need access to GitHub secrets. If you need this work in the `dev` branch ASAP @arengel, you are listed as a collaborator on the PUDL repo so you can create a new PR from a branch inside this repo so the actions will work.

Just created #2251 which is the same content as this but should be easier since it is from PUDL proper. I had been having permission issues pushing to PUDL and finally figured it out. Merging it in is not super urgent but I imagine the revised PR is generally easier to deal with so closing this in favor of the new one.
